src/main/java/net/dries007/tfc/common/blockentities/BarrelBlockEntity.java

```diff
@@ -0,0 +1,478 @@
+        }
+
+        final SealedBarrelRecipe recipe = barrel.recipe;
+        if (recipe != null && state.getValue(BarrelBlock.SEALED))
+        {
+            final int durationSealed = (int) (Calendars.SERVER.getTicks() - barrel.recipeTick);
+            if (!recipe.isInfinite() && durationSealed > recipe.getDuration())
+            {
+                if (recipe.matches(barrel.inventory, level))
+                {
+                    // Recipe completed, so fill outputs
+                    recipe.assembleOutputs(barrel.inventory);
+                    Helpers.playSound(level, barrel.getBlockPos(), recipe.getCompleteSound());
+                }
+
+                // In both cases, update the recipe and sync
+                barrel.updateRecipe();
+                barrel.markForSync();
+            }
+        }
+
+        if (barrel.needsInstantRecipeUpdate)
+        {
+            barrel.needsInstantRecipeUpdate = false;
+            if (barrel.inventory.excess.isEmpty()) // Excess must be empty for instant recipes to apply
+            {
+                level.getRecipeManager().getRecipeFor(TFCRecipeTypes.BARREL_INSTANT.get(), barrel.inventory, level)
+                    .ifPresent(instantRecipe -> {
+                        instantRecipe.assembleOutputs(barrel.inventory);
+                        Helpers.playSound(level, barrel.getBlockPos(), instantRecipe.getCompleteSound());
+                    });
+                barrel.markForSync();
+            }
+        }
+    }
+
+    private final SidedHandler.Builder<IFluidHandler> sidedFluidInventory;
+    private final IntArrayBuilder syncableData;
+
+    @Nullable private SealedBarrelRecipe recipe;
+    private long lastUpdateTick; // The last tick this barrel was updated in serverTick()
+    private long sealedTick; // The tick this barrel was sealed
+    private long recipeTick; // The tick this barrel started working on the current recipe
+
+    private boolean needsRecipeUpdate;
+    private boolean needsInstantRecipeUpdate; // If the instant recipe needs to be checked again
+
+    public BarrelBlockEntity(BlockPos pos, BlockState state)
+    {
+        super(TFCBlockEntities.BARREL.get(), pos, state, BarrelInventory::new, NAME);
+
+        sidedInventory
+            .on(new PartialItemHandler(inventory).insert(SLOT_FLUID_CONTAINER_IN).extract(SLOT_FLUID_CONTAINER_OUT), Direction.Plane.HORIZONTAL)
+            .on(new PartialItemHandler(inventory).insert(SLOT_ITEM), Direction.UP)
+            .on(new PartialItemHandler(inventory).extract(SLOT_ITEM), Direction.DOWN);
+
+        sidedFluidInventory = new SidedHandler.Builder<IFluidHandler>(inventory)
+            .on(new PartialFluidHandler(inventory).insert(), Direction.UP)
+            .on(new PartialFluidHandler(inventory).extract(), Direction.DOWN, Direction.NORTH, Direction.SOUTH, Direction.EAST, Direction.WEST);
+
+        syncableData = new IntArrayBuilder();
```

Comment:
I think I originally added this just in copy-paste "this will probably be needed", but looking at this, it seems not, so we could feel free to remove it. (Unless it has some declarations I am missing)